File: cpp/48. Rotate Image.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;

class Solution {
public:
    void rotate(vector<vector<int>>& matrix) {
        int m = matrix.size(), n = matrix[0].size();
        for(int i = 0; i < m; i++) {
            for(int j = i + 1; j < n; j++) {
                swap(matrix[i][j], matrix[j][i]);
            }
        }
        for(auto &row : matrix) {
            reverse(row.begin(), row.end());
        }
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    Solution solution; vvi matrix;

    solution = Solution();
    matrix = {{1,2,3},{4,5,6},{7,8,9}};
    solution.rotate(matrix);
    assert(matrix==vvi({{7,4,1},{8,5,2},{9,6,3}}));

    solution = Solution();
    matrix = {{5,1,9,11},{2,4,8,10},{13,3,6,7},{15,14,12,16}};
    solution.rotate(matrix);
    assert(matrix==vvi({{15,13,2,5},{14,3,4,1},{12,6,8,9},{16,7,10,11}}));
}

File: python/48-rotate-image.py
from typing import List

'''
1. Transpose the matrix
2. reverse each row

ex: 
matrix = [[1,2,3], 
          [4,5,6],
          [7,8,9]]
transpose -> [[1,4,7], 
              [2,5,8],
              [3,6,9]]
reverse each row -> [[7,4,1], 
                     [8,5,2],
                     [9,6,3]]
'''
class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        "Time: O(n^2), Space: O(1)"
        # Transpose
        for i in range(len(matrix)):
            for j in range(i+1, len(matrix[i])):
                matrix[i][j], matrix[j][i] = matrix[j][i], matrix[i][j]
        # Reverse each row
        for row in matrix:
            row.reverse()


matrix = [[1,2,3], 
          [4,5,6],
          [7,8,9]]
Solution().rotate(matrix)
assert matrix == [[7,4,1],
                  [8,5,2],
                  [9,6,3]]

matrix = [[5,1,9,11],
          [2,4,8,10],
          [13,3,6,7],
          [15,14,12,16]]
Solution().rotate(matrix)
assert matrix == [[15,13,2,5],
                  [14,3,4,1],
                  [12,6,8,9],
                  [16,7,10,11]]

matrix = [[1]]
Solution().rotate(matrix)
assert matrix == [[1]]

matrix = [[1,2],
          [3,4]]
Solution().rotate(matrix)
assert matrix == [[3,1],
                  [4,2]]